relayer/relays/execution/message/message.go

```diff
@@ -0,0 +1,320 @@
+package message
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common"
+	log "github.com/sirupsen/logrus"
+	"github.com/snowfork/snowbridge/relayer/chain"
+	"github.com/snowfork/snowbridge/relayer/chain/ethereum"
+	"github.com/snowfork/snowbridge/relayer/chain/parachain"
+	"github.com/snowfork/snowbridge/relayer/relays/execution/config"
+	"golang.org/x/sync/errgroup"
+)
+
+type Message struct {
+	writer    *parachain.ParachainWriter
+	listener  *EthereumListener
+	addresses []common.Address
+}
+
+func New(ctx context.Context, eg *errgroup.Group, writer *parachain.ParachainWriter, configSource *config.SourceConfig, ethconn *ethereum.Connection) (*Message, error) {
+	addresses, err := configSource.GetBasicChannelAddresses()
+	if err != nil {
+		return nil, err
+	}
+
+	listener := NewEthereumListener(
+		configSource,
+		ethconn,
+	)
+
+	err = listener.Start(ctx, eg)
+	if err != nil {
+		return &Message{}, err
+	}
+
+	return &Message{writer, listener, addresses}, nil
+}
+
+func (m *Message) Sync(ctx context.Context, eg *errgroup.Group) error {
+	lastSyncedIncentivizedBlock, err := m.syncUnprocessedIncentivizedMessages(ctx)
+	if err != nil {
+		return fmt.Errorf("sync unprocessed incentivized messages: %w", err)
+	}
+
+	lastSyncedBasicBlock, err := m.syncUnprocessedBasicMessages(ctx)
+	if err != nil {
+		return fmt.Errorf("sync unprocessed basic messages: %w", err)
+	}
+
+	ticker := time.NewTicker(time.Second * 20)
+
+	var secondLastSyncedBlockNumber uint64
+	if lastSyncedIncentivizedBlock > lastSyncedBasicBlock {
+		secondLastSyncedBlockNumber = lastSyncedIncentivizedBlock
+	} else {
+		secondLastSyncedBlockNumber = lastSyncedBasicBlock
```

Comment:
Why is `secondLastSyncedBlockNumber` the max of the last synced blocks for each channel? If syncing fails for one channel and not the other for one block, we could ignore some blocks for the channel that was failing. If that block contains messages, a skipped nonce could prevent the channel from receiving future messages.

eg.
```
lastSyncedIncentivizedBlock = 11
lastSyncedBasicBlock = 10
secondLastSyncedBlockNumber = 11
lastSyncedBlockNumber = 12
```

Then we'll try to sync block 12 for both channels, but if block 11 contains a basic channel message, the channel will reject it because the nonce isn't incrementing by 1.

Given that syncing can fail independently (basic channel sync could fail after incentivized channel sync succeeds), could we use `lastSyncedIncentivizedBlock` and `lastSyncedBasicBlock` separately? `secondLastSyncedBlockNumber` would still be useful in the loop to detect unchanged execution blocks, but we'd need to start with the min of `lastSyncedIncentivizedBlock` and `lastSyncedBasicBlock` here.

Reply:
Good point! I thought it would be OK because blocks are always processed sequentially, but realized that is not actually the case if one channel gets processed before the other and the other channel has an earlier block with a message. Fixed in https://github.com/Snowfork/snowbridge/pull/721/commits/dc0ef6af10d04299d197a92a6fb564d5f086448b. The 2 3 variables assigned to `lastSyncedBlockNumber` isn't the best, I'll see if I can refactor that a bit.

Reply:
I ended up removing `secondLastSyncedBlockNumber ` here: https://github.com/Snowfork/snowbridge/pull/721/commits/e4b51ee9109baff6ffcfaddaaf5de0f7d83a76ec